File: backend/app/api/packages.py
"""Packages API — P0 阶段的 registry / manifest 校验 / 安装管理接口。"""

from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field, ValidationError

from app.models.database import async_session
from app.services.plugin_registry import (
    compare_registry_package_versions,
    get_registry_package,
    import_plugin_source,
    import_plugin_source_spec,
    install_registry_package,
    list_artifact_variants,
    list_execution_logs,
    list_installed_packages,
    list_registry_packages,
    list_registry_package_versions,
    list_workflow_bindings,
    rollback_installed_package,
    toggle_installed_package,
    upgrade_installed_package,
    validate_manifest_payload,
)
from app.services.remote_package_sources import create_github_remote_source, infer_package_id

router = APIRouter(prefix="/packages", tags=["packages"])

DEFAULT_USER_ID = "default-user-00000000"


class PackageInstallRequest(BaseModel):
    package_id: str = Field(..., min_length=3, max_length=127)
    version: str | None = Field(default=None, min_length=5, max_length=30)


class PackageUpgradeRequest(BaseModel):
    target_version: str | None = Field(default=None, min_length=5, max_length=30)


class PackageToggleRequest(BaseModel):
    enabled: bool


class PackageImportRequest(BaseModel):
    source_id: str | None = Field(default=None, min_length=3, max_length=127)
    owner: str | None = Field(default=None, min_length=1, max_length=127)
    repo: str | None = Field(default=None, min_length=1, max_length=127)
    ref: str = Field(default="main", min_length=1, max_length=127)
    plugin_path: str | None = Field(default=None, min_length=1, max_length=255)
    package_id: str | None = Field(default=None, min_length=3, max_length=127)
    package_kind: Literal["workflow", "tool_adapter"] = "workflow"
    related_skill_path: str | None = Field(default=None, min_length=1, max_length=255)
    adapter_targets: list[str] = Field(default_factory=list)


class ManifestValidationRequest(BaseModel):
    manifest: dict


@router.get("/registry")
async def get_registry():
    async with async_session() as session:
        items = await list_registry_packages(session)
        return {"items": items, "total": len(items)}


@router.get("/registry/{package_id}")
async def get_registry_item(package_id: str):
    async with async_session() as session:
        item = await get_registry_package(session, package_id)
        if item is None:
            raise HTTPException(status_code=404, detail="Package 不存在")
        return {"item": item.model_dump()}


@router.get("/registry/{package_id}/versions")
async def get_registry_item_versions(package_id: str):
    async with async_session() as session:
        items = await list_registry_package_versions(session, package_id)
        if not items:
            raise HTTPException(status_code=404, detail="Package 不存在")
        return {
            "package_id": package_id,
            "latest_version": items[0]["version"],
            "versions": items,
        }


@router.post("/validate-manifest")
async def validate_manifest(data: ManifestValidationRequest):
    try:
        manifest = validate_manifest_payload(data.manifest)
    except (ValueError, ValidationError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {"manifest": manifest, "valid": True}


@router.get("/installed")
async def get_installed():
    async with async_session() as session:
        items = await list_installed_packages(session, DEFAULT_USER_ID)
        return {"items": items, "total": len(items)}


@router.post("/import")
async def import_package_source(data: PackageImportRequest):
    async with async_session() as session:
        try:
            if data.source_id:
                result = await import_plugin_source(session, DEFAULT_USER_ID, data.source_id)
            else:
                if not data.owner or not data.repo or not data.plugin_path:
                    raise HTTPException(
                        status_code=400,
                        detail="自定义 GitHub 导入至少需要 owner、repo 和 plugin_path",
                    )
                package_id = data.package_id or infer_package_id(data.owner, data.repo, data.plugin_path)
                spec = create_github_remote_source(
                    owner=data.owner,
                    repo=data.repo,
                    ref=data.ref,
                    plugin_path=data.plugin_path,
                    package_id=package_id,
                    package_kind=data.package_kind,
                    related_skill_path=data.related_skill_path,
                    adapter_targets=tuple(target.strip() for target in data.adapter_targets if target.strip()),
                )
                result = await import_plugin_source_spec(session, DEFAULT_USER_ID, spec)
        except ValueError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except RuntimeError as exc:
            raise HTTPException(status_code=502, detail=str(exc)) from exc
        return result


@router.post("/install")
async def install_package(data: PackageInstallRequest):
    async with async_session() as session:
        try:
            items = await install_registry_package(
                session,
                DEFAULT_USER_ID,
                data.package_id,
                version=data.version,
            )
        except ValueError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        return {"installed_packages": items, "total": len(items)}


@router.get("/{package_id}/compare")
async def compare_package_versions(
    package_id: str,
    from_version: str,
    to_version: str,
):
    async with async_session() as session:
        try:
            result = await compare_registry_package_versions(session, package_id, from_version, to_version)
        except ValueError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        return result


@router.post("/{package_id}/upgrade")
async def upgrade_package(package_id: str, data: PackageUpgradeRequest):
    async with async_session() as session:
        try:
            items = await upgrade_installed_package(
                session,
                DEFAULT_USER_ID,
                package_id,
                target_version=data.target_version,
            )
        except LookupError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"updated_packages": items, "total": len(items)}


@router.post("/{package_id}/rollback")
async def rollback_package(package_id: str):
    async with async_session() as session:
        try:
            items = await rollback_installed_package(session, DEFAULT_USER_ID, package_id)
        except LookupError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"updated_packages": items, "total": len(items)}


@router.post("/{package_id}/toggle")
async def toggle_package(package_id: str, data: PackageToggleRequest):
    async with async_session() as session:
        item = await toggle_installed_package(session, DEFAULT_USER_ID, package_id, data.enabled)
        if item is None:
            raise HTTPException(status_code=404, detail="Package 未安装")
        return {"item": item}


@router.get("/workflow-bindings")
async def get_workflow_bindings(
    package_id: str | None = None,
    presentation_id: str | None = None,
    task_id: str | None = None,
):
    async with async_session() as session:
        items = await list_workflow_bindings(
            session,
            DEFAULT_USER_ID,
            package_id=package_id,
            presentation_id=presentation_id,
            task_id=task_id,
        )
        return {"items": items, "total": len(items)}


@router.get("/execution-logs")
async def get_package_execution_logs(
    package_id: str | None = None,
    status: str | None = None,
    limit: int = 50,
):
    async with async_session() as session:
        items = await list_execution_logs(
            session,
            DEFAULT_USER_ID,
            package_id=package_id,
            status=status,
            limit=limit,
        )
        return {"items": items, "total": len(items)}


@router.get("/artifact-variants")
async def get_package_artifact_variants(
    package_id: str | None = None,
    presentation_id: str | None = None,
    asset_id: str | None = None,
):
    async with async_session() as session:
        items = await list_artifact_variants(
            session,
            DEFAULT_USER_ID,
            package_id=package_id,
            presentation_id=presentation_id,
            asset_id=asset_id,
        )
        return {"items": items, "total": len(items)}

File: backend/data/uploads/_projects/7be8acb22e1c/presentationagent/frontend/src/lib/packages.ts
export type PackageKind = "foundation" | "workflow" | "skill" | "theme" | "tool_adapter";

export interface PackagePermission {
  name: string;
  rationale: string;
}

export interface PackageDependency {
  package_id: string;
  version_constraint: string;
  optional?: boolean;
}

export interface PackageEntrypoint {
  kind: "workflow" | "skill_set" | "theme_bundle" | "adapter";
  target: string;
  description: string;
}

export interface PackageCompatibility {
  min_platform_version: string;
  target_artifact_mode: Array<"native_pptx_first" | "dual_render">;
}

export interface InstalledPackageHistoryEntry {
  version: string;
  changed_at?: string | null;
  action: string;
}

export interface PackageVersionSummary {
  version: string;
  display_name: string;
  description: string;
  release_notes: string;
  release_date?: string | null;
  is_latest: boolean;
  capability_count: number;
  permission_count: number;
  dependency_count: number;
}

export interface PackageVersionCompareResult {
  package_id: string;
  from_version: string;
  to_version: string;
  direction: "upgrade" | "rollback" | "same";
  from_manifest: PackageManifest;
  to_manifest: PackageManifest;
  added_capabilities: string[];
  removed_capabilities: string[];
  added_permissions: PackagePermission[];
  removed_permissions: PackagePermission[];
  added_dependencies: PackageDependency[];
  removed_dependencies: PackageDependency[];
  release_notes?: string;
  upgrade_notes?: string;
}

export interface PackageManifest {
  schema_version: string;
  package_id: string;
  display_name: string;
  kind: PackageKind;
  version: string;
  description: string;
  publisher: string;
  tags: string[];
  capabilities: string[];
  permissions: PackagePermission[];
  dependencies: PackageDependency[];
  compatibility: PackageCompatibility;
  entrypoints: PackageEntrypoint[];
  metadata?: Record<string, string>;
}

export interface PackageImportResult {
  source_id: string;
  source_ref?: string;
  package_ids: string[];
  versions: string[];
  latest_manifest?: PackageManifest | null;
}

export interface PackageImportRequest {
  source_id?: string;
  owner?: string;
  repo?: string;
  ref?: string;
  plugin_path?: string;
  package_id?: string;
  package_kind?: "workflow" | "tool_adapter";
  related_skill_path?: string;
  adapter_targets?: string[];
}

export interface ExportFormatCapability {
  available: boolean;
  reason?: string | null;
}

export interface ExportCapabilitiesResponse {
  formats: Record<string, ExportFormatCapability>;
}

export interface PackageExecutionLogRecord {
  id: string;
  package_id: string;
  package_version?: string | null;
  execution_kind: string;
  target_type?: string | null;
  target_id?: string | null;
  status: string;
  duration_ms?: number | null;
  error_message?: string | null;
  input_payload?: Record<string, unknown> | null;
  output_payload?: Record<string, unknown> | null;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface ArtifactVariantRecord {
  id: string;
  variant_key: string;
  package_id: string;
  package_version?: string | null;
  variant_type: string;
  presentation_id?: string | null;
  asset_id?: string | null;
  file_url?: string | null;
  mime_type?: string | null;
  metadata?: Record<string, unknown> | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface InstalledPackageRecord {
  id: string;
  package_id: string;
  display_name: string;
  package_kind: PackageKind;
  version: string;
  source: string;
  manifest: PackageManifest;
  granted_permissions: PackagePermission[];
  status: string;
  is_enabled: boolean;
  installed_at?: string | null;
  updated_at?: string | null;
  latest_version?: string | null;
  previous_version?: string | null;
  available_versions?: string[];
  upgrade_available?: boolean;
  installed_history?: InstalledPackageHistoryEntry[];
  release_notes?: string | null;
  latest_release_notes?: string | null;
}

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init);
  if (!response.ok) {
    let detail = `HTTP ${response.status}`;
    try {
      const data = await response.json();
      detail = data.detail || data.message || detail;
    } catch {
      // ignore json parse failures
    }
    throw new Error(detail);
  }
  return response.json() as Promise<T>;
}

export async function getRegistryPackages(): Promise<PackageManifest[]> {
  const data = await fetchJson<{ items: PackageManifest[] }>("/api/packages/registry");
  return data.items || [];
}

export async function getRegistryPackageVersions(packageId: string): Promise<PackageVersionSummary[]> {
  const data = await fetchJson<{ versions: PackageVersionSummary[] }>(`/api/packages/registry/${packageId}/versions`);
  return data.versions || [];
}

export async function getInstalledPackages(): Promise<InstalledPackageRecord[]> {
  const data = await fetchJson<{ items: InstalledPackageRecord[] }>("/api/packages/installed");
  return data.items || [];
}

export async function installRegistryPackage(
  packageId: string,
  version?: string,
): Promise<InstalledPackageRecord[]> {
  const data = await fetchJson<{ installed_packages: InstalledPackageRecord[] }>("/api/packages/install", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ package_id: packageId, version }),
  });
  return data.installed_packages || [];
}

export async function importPackageSource(request: string | PackageImportRequest): Promise<PackageImportResult> {
  const payload = typeof request === "string" ? { source_id: request } : request;
  return fetchJson<PackageImportResult>("/api/packages/import", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function upgradeInstalledPackage(
  packageId: string,
  targetVersion?: string,
): Promise<InstalledPackageRecord[]> {
  const data = await fetchJson<{ updated_packages: InstalledPackageRecord[] }>(`/api/packages/${packageId}/upgrade`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ target_version: targetVersion }),
  });
  return data.updated_packages || [];
}

export async function rollbackInstalledPackage(packageId: string): Promise<InstalledPackageRecord[]> {
  const data = await fetchJson<{ updated_packages: InstalledPackageRecord[] }>(`/api/packages/${packageId}/rollback`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({}),
  });
  return data.updated_packages || [];
}

export async function comparePackageVersions(
  packageId: string,
  fromVersion: string,
  toVersion: string,
): Promise<PackageVersionCompareResult> {
  const params = new URLSearchParams({ from_version: fromVersion, to_version: toVersion });
  return fetchJson<PackageVersionCompareResult>(`/api/packages/${packageId}/compare?${params.toString()}`);
}

export async function toggleInstalledPackage(
  packageId: string,
  enabled: boolean,
): Promise<InstalledPackageRecord> {
  const data = await fetchJson<{ item: InstalledPackageRecord }>(`/api/packages/${packageId}/toggle`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ enabled }),
  });
  return data.item;
}

export async function getPackageExecutionLogs(
  packageId: string,
  options?: { status?: string; limit?: number },
): Promise<PackageExecutionLogRecord[]> {
  const params = new URLSearchParams({ package_id: packageId });
  if (options?.status) params.set("status", options.status);
  if (options?.limit) params.set("limit", String(options.limit));
  const data = await fetchJson<{ items: PackageExecutionLogRecord[] }>(
    `/api/packages/execution-logs?${params.toString()}`,
  );
  return data.items || [];
}

export async function getPackageArtifactVariants(
  packageId: string,
  options?: { presentationId?: string; assetId?: string },
): Promise<ArtifactVariantRecord[]> {
  const params = new URLSearchParams({ package_id: packageId });
  if (options?.presentationId) params.set("presentation_id", options.presentationId);
  if (options?.assetId) params.set("asset_id", options.assetId);
  const data = await fetchJson<{ items: ArtifactVariantRecord[] }>(
    `/api/packages/artifact-variants?${params.toString()}`,
  );
  return data.items || [];
}

export function getPackageKindLabel(kind: PackageKind): string {
  const labels: Record<PackageKind, string> = {
    foundation: "基础契约",
    workflow: "工作流",
    skill: "Skill 包",
    theme: "主题包",
    tool_adapter: "工具适配器",
  };
  return labels[kind] || kind;
}

export function getPermissionLabel(name: string): string {
  const labels: Record<string, string> = {
    "asset.read": "读公共空间",
    "asset.write": "写公共空间",
    "document.parse": "解析文档",
    "model.invoke": "调用模型",
    "pptx.render": "渲染 PPTX",
    "preview.render": "渲染预览",
    "registry.read": "读注册表",
    "settings.write": "改设置",
    "web.fetch": "联网访问",
  };
  return labels[name] || name;
}

export function getPackageStatusLabel(item: InstalledPackageRecord): string {
  if (!item.is_enabled) return "已禁用";
  if (item.status === "rolled_back") return "已回滚";
  if (item.status === "upgraded") return "已升级";
  return "已启用";
}

export function getPackageStatusTone(item: InstalledPackageRecord): string {
  if (!item.is_enabled) return "bg-gray-100 text-gray-600";
  if (item.status === "rolled_back") return "bg-amber-50 text-amber-700";
  if (item.status === "upgraded") return "bg-blue-50 text-blue-700";
  return "bg-green-50 text-green-700";
}

export function matchesPackageSearch(manifest: PackageManifest, searchTerm: string): boolean {
  const keyword = searchTerm.trim().toLowerCase();
  if (!keyword) return true;

  const haystacks = [
    manifest.package_id,
    manifest.display_name,
    manifest.description,
    manifest.publisher,
    ...manifest.tags,
    ...manifest.capabilities,
  ];
  return haystacks.some((value) => value?.toLowerCase().includes(keyword));
}

export function normalizePackageFileUrl(url?: string | null): string | null {
  if (!url) return null;
  if (/^https?:\/\//i.test(url)) return url;
  if (url.startsWith("/")) return url;
  return `/${url}`;
}